spring-cloud-starter-stream-source-rabbit/src/test/java/org/springframework/cloud/stream/app/rabbit/source/RabbitSourceInvalidConfigTests.java

```diff
@@ -33,39 +36,49 @@
  * Tests for RabbitSource with invalid config.
  *
  * @author Gary Russell
+ * @author Chris Schaefer
  */
 public class RabbitSourceInvalidConfigTests {
 
 	@Test
 	public void testNoQueues() throws Exception {
 		try {
 			AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
-			EnvironmentTestUtils.addEnvironment(context, "rabbit.enableRetry:false");
+			TestPropertyValues.of("rabbit.enableRetry:false").applyTo(context);
 			context.register(Config.class);
 			context.refresh();
 			fail("BeanCreationException expected");
 		}
 		catch (Exception e) {
 			assertThat(e, instanceOf(BeanCreationException.class));
-			assertThat(e.getMessage(), containsString("queue(s) are required"));
+			assertThat(extractedValidationMessage(e), containsString("queue(s) are required"));
 		}
 	}
 
 	@Test
 	public void testEmptyQueues() throws Exception {
 		try {
 			AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext();
-			EnvironmentTestUtils.addEnvironment(context, "rabbit.enableRetry:false", "rabbit.queues:");
+			TestPropertyValues.of("rabbit.enableRetry:false").applyTo(context);
+			TestPropertyValues.of("rabbit.queues:").applyTo(context);
 			context.register(Config.class);
 			context.refresh();
 			fail("BeanCreationException expected");
 		}
 		catch (Exception e) {
 			assertThat(e, instanceOf(BeanCreationException.class));
-			assertThat(e.getMessage(), containsString("At least one queue is required"));
+			assertThat(extractedValidationMessage(e), containsString("At least one queue is required"));
 		}
 	}
 
+	private String extractedValidationMessage(Exception e) {
```

Comment:
This method sounds like a utility component to pulled to the app-strarters-core level: https://github.com/spring-cloud-stream-app-starters/core

I've seen this before but was too lazy to try to extract the target validation error.
IMHO the previous version of the `BindValidationException` was better.
Maybe worth to ask from Spring Boot team to reinstate?

Reply:
i was hoping https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/diagnostics/analyzer/BindValidationFailureAnalyzer.java would have ben usable, but according to https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/diagnostics/analyzer/package-info.java its indicated as an "internal" implementation.

my method is pretty basic and may not cover all use cases -- would be great to get the BindValidationFailureAnalyzer opened up as it *appears* it would provide similar concept rather than re-invent as a util in app-starters-core

what was the "previous version of the `BindValidationException` your mentioning?

Reply:
Great! Let's ping Boot team to make that class `public`!

Well, with the previous Boot version we didn't need any additional tricks to get the target validation error. 
See the code you have just changed from the simple `e.getMessage(), containsString()` to this `extractedValidationMessage()` utility.

Reply:
ok - i like the idea of asking the boot team if that can be opened up or maybe there is something similar. relying on parsing logged messages seems a bit fragile. will pickup next week

Reply:
since you mentioned making a utility class out of this, do you feel this type of validation is a common occurrence across app starters? its the first time i have encountered it, so curious as i had received a reply "I’m not aware of any utilities. The failure analyzer is intentionally package-private as we don’t want it to be subject to the restrictions that come with being public API. Feel free to copy any of the code that’s useful to you though".

id feel a bit awkward basically doing a copy/paste on that class into our project (its more robust/utility class suited than my one off impl here). if this is a common validation pattern i suppose its fine to to do and put a note in the javadoc. but if not, digging down into the stack on a here and there basis doesn't seem too complicated (and might be different per situation).

Reply:
I think we can merge this as is and move the discussion about the top level utility to some other place.

@wilkinsona,

give us, please, some conclusion from your side how should we proceed or you will do the fix on Boot level in the end anyway. According to our private discussion in Slack.
Thanks

Reply:
works for me 👍

Reply:
Quoting Andy:

> Andy Wilkinson 
I’d write your own for now. I haven’t heard of anyone else needing it

>Andy Wilkinson [< 1 minute ago]
It’d be a reasonably big piece of work to write something that’s general purpose

So, @chrisjs , we have to go with our own util in the parent project for all the starters or just resolve issues from case to case.